### DISCUSSION TITLE
求教问题出在哪，ArrayField列表渲染超过十条之后开始卡顿

### DISCUSSION THREAD

**Discussion by sogud** [alibaba/formily#2695] [Q&A 2.x] [answered]
https://codesandbox.io/s/happy-johnson-vmigg?file=/RuleGroupItem.tsx

**Comment by janryWang** [comment, marked as answer]
看了之后，发现其实问题就这几点，优化的好的话，分分钟丝滑般体验

1. 为啥要把SchemaField传给ObjectField使用，这种用法好奇怪，还套了一层又一层，SchemaField它本身就是个Field，建议把SchemaField单独使用
2. 每个维度你嵌套了很多Field组件，一个Field组件渲染差不多要1ms，你大概嵌套了10个，如果按照10行算，那就是100ms，每次点击新增都会有100ms卡顿，建议减少嵌套层级
3. ConditionItem接收的属性是一个Field模型，Field是一个很庞大的数据结构，传到组件属性上去的话，每次都会递归toJS去依赖收集整树，所以，非常不推荐在componentProps上传复杂属性，最佳实践是通过react context传递

**Comment by janryWang** [comment]
还有就是，需要注意一点，自定义组件内如果用到FormLayout FormGrid这种布局组件，是可以直接使用的，因为它就是普通React组件，不需要再用VoidField来代理状态，这样就能减少实际化Field的成本了

**Comment by sogud** [comment]
根据大佬的解答确实找到问题，是因为原因3，组件接收的属性是一个Field模型导致toJS性能降低，不通过props传入就好了，可以使用useField query方式获取父级FIeld模型。
